neps/nep-0452.md

```diff
@@ -0,0 +1,164 @@
+
+## Specification
+
+```ts
+/// Information about a specific public key.
+type KeyInfo = {
+   /// yoctoNEAR$ amount that will be sent to the claiming account (either new or existing)
+   // when the key is successfully used.
+   balance: string,
+
+   /// If using the NFT standard extension, a set of NFTData can be linked to the public key      
+   /// indicating that all those assets will be sent to the claiming account (either new or   
+   /// existing) when the key is successfully used.
+   nft_data: Vec<NFTData> | null,
+  
+   /// If using the FT standard extension, a set of FTData can be linked to the public key      
+   /// indicating that all those assets will be sent to the claiming account (either new or   
+   /// existing) when the key is successfully used.
+   ft_data: Vec<FTData> | null,
+
+   // ... other types can be introduced and the standard is easily extendable.
+}
+
+
+/// Data outlining a specific Non-Fungible Token that should be sent to the claiming account /// (either new or existing) when a key is successfully used.
+type NFTData = {
+   /// the id of the token to transfer
+   token_id: string,
+   /// The valid NEAR account indicating the Non-Fungible Token contract.
+   contract_id: string
+}
+
+
+/// Data outlining Fungible Tokens that should be sent to the claiming account 
+/// (either new or existing) when a key is successfully used.
+type FTData = {
+   /// The number of tokens to transfer, wrapped in quotes and treated
+   /// like a string, although the number will be stored as an unsigned integer
+   /// with 128 bits.
+   amount: string,
+   /// The valid NEAR account indicating the Fungible Token contract.
+   contract_id: string
+}
+
+
+
+
+/****************/
+/* VIEW METHODS */
+/****************/
+
+// Returns the $yoctoNEAR amount associated with a given public key 
+// Panics if the key does not exist 
+function get_key_balance(key: PublicKey) -> string
+
+// Returns the KeyInfo associated with a given public key
+// Panics if the key does not exist
+function get_key_information(key: PublicKey) -> KeyInfo
+
+/******************/
+/* CHANGE METHODS */
+/******************/
+
+// Transfer all assets linked to the signer’s public key to an *existing* NEAR account. 
+// If the transfer fails for whatever reason, it is up to the smart contract developer to 
+// choose what should happen. For example, the contract can choose to keep the assets 
+// or send them back to the original linkdrop creator.
+// 
+// Requirements:
+// * The predecessor account *MUST* be the current contract ID.
+// * The assets being sent *MUST* be associated with the signer’s public key.
+// * The assets *MUST* be sent to the `account_id` passed in.
+//
+// Arguments:
+// * `account_id` the account that should receive the linkdrop assets.
+function claim(account_id: string) -> Promise
```

Comment:
Given that we use `PublicKey` type above, should we consider using `AccountId` type instead of a string here?

Reply:
Good point, I think we should stray away from using rust SDK specific data types and keep it generic `i.e string`

Reply:
I would rather turn it the other way around and define `AccountId` and `PublicKey` types and reference them everywhere.

Reply:
I was trying to follow how it's done in other standards. For example, in the NFT standard, there is no mention of Account ID or rust SDK specific types. It is all just `string`, `number` etc.

Reply:
Yeah, but as time goes, there is more and more repetition in what "account id", "u128 string", and "public key" is, so I think we should consider some template with the commonly used types which we can embed in the NEPs going forward. These are just my thoughts out loud, no action items for you 😄